### PR TITLE
feat(mask): add support for image file bitmap masks

### PR DESCRIPTION
**Pull request by kisvegabor**
### Description of the feature or fix

cc #XuNeo

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: src/draw/sw/lv_draw_sw_img.c [hunk 1]
```diff
@@ -236,22 +236,35 @@ static void img_draw_core(lv_draw_unit_t * draw_unit, const lv_draw_image_dsc_t
     }
     /*Handle masked RGB565, RGB888, XRGB888, or ARGB8888 images*/
     else if(!transformed && masked && draw_dsc->recolor_opa <= LV_OPA_MIN) {
+        lv_image_decoder_dsc_t mask_decoder_dsc;
+        lv_area_t mask_area;
+        lv_result_t decoder_res = lv_image_decoder_open(&mask_decoder_dsc, draw_dsc->bitmap_mask_src, NULL);
+        if(decoder_res == LV_RESULT_OK && mask_decoder_dsc.decoded) {
+            const lv_draw_buf_t * mask_img = mask_decoder_dsc.decoded;
+            blend_dsc.mask_buf = mask_img->data;
```
Comment (XuNeo) [inline src/draw/sw/lv_draw_sw_img.c]: Let's check if the image color format is A8 and ignore it here.

In future, we can add decoder args "as_mask" etc. to let decoder convert image to A8 format.

File: tests/src/test_cases/draw/test_draw_layer.c [hunk 2]
```diff
@@ -31,7 +31,42 @@ void test_draw_layer_bitmap_mask(void)
...
+    lv_obj_set_style_bitmap_mask_src(obj, "A:src/test_files/binimages/cogwheel.A8.bin", 0);
+    lv_obj_center(obj);
+
+    lv_obj_t * label = lv_label_create(obj);
+    lv_obj_set_width(label, lv_pct(100));
+    lv_label_set_text(label,
+                      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque suscipit risus nec pharetra pulvinar. In hac habitasse platea dictumst. Proin placerat congue massa eu luctus. Suspendisse risus nulla, consectetur eget odio ut, mollis sollicitudin magna. Suspendisse volutpat consequat laoreet. Aenean sodales suscipit leo, vitae pulvinar lorem pulvinar eu. Nullam molestie hendrerit est sit amet imperdiet.");
+    lv_obj_center(label);
+
+    TEST_ASSERT_EQUAL_SCREENSHOT("draw/draw_layer_bitmap_mask.png");
```
Comment (XuNeo) [inline tests/src/test_cases/draw/test_draw_layer.c:50]: You may forget uploading the reference image.

File: tests/src/test_cases/draw/test_draw_layer.c [hunk 3]
```diff
@@ -31,7 +31,42 @@ void test_draw_layer_bitmap_mask(void)
...
+    lv_obj_set_style_bitmap_mask_src(obj, "A:not/exisiting/file", 0);
+    lv_obj_center(obj);
 
+    lv_obj_t * label = lv_label_create(obj);
+    lv_obj_set_width(label, lv_pct(100));
+    lv_label_set_text(label,
+                      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque suscipit risus nec pharetra pulvinar. In hac habitasse platea dictumst. Proin placerat congue massa eu luctus. Suspendisse risus nulla, consectetur eget odio ut, mollis sollicitudin magna. Suspendisse volutpat consequat laoreet. Aenean sodales suscipit leo, vitae pulvinar lorem pulvinar eu. Nullam molestie hendrerit est sit amet imperdiet.");
+    lv_obj_center(label);
+
+    TEST_ASSERT_EQUAL_SCREENSHOT("draw/draw_layer_bitmap_mask_not_masked.png");
```
Comment (XuNeo) [inline tests/src/test_cases/draw/test_draw_layer.c:68]: ditto.

**Review by XuNeo — COMMENTED** [review]
Thank you Gabor. That's really quick!

File: tests/src/test_cases/draw/test_draw_layer.c (hunk 2 above)
Comment (kisvegabor) [reply tests/src/test_cases/draw/test_draw_layer.c:50]: It's using an already existing image. See [here](https://github.com/lvgl/lvgl/blob/master/tests/ref_imgs/draw/draw_layer_bitmap_mask.png).

File: src/draw/sw/lv_draw_sw_img.c (hunk 1 above)
Comment (kisvegabor) [reply src/draw/sw/lv_draw_sw_img.c]: Fair point, added.

File: tests/src/test_cases/draw/test_draw_layer.c (hunk 3 above)
Comment (kisvegabor) [reply tests/src/test_cases/draw/test_draw_layer.c:68]: Added.

**Review by XuNeo — DISMISSED** [review]

**Review by XuNeo — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**